### PR TITLE
Fix collapsed InputDecorator minimum height

### LINKED ISSUE #150763: Input text does not fit into a collapsed InputDecoration on desktop
### Steps to reproduce

1. Run the code sample on desktop

### Expected results

The text fits in the decoration

image

### Actual results

The decoration is too small (letters 'p' and 'g' overflows the background, underline border is misplaced)

image

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() async {
  runApp(
    const MaterialApp(
      title: 'Flutter Demo',
      home: MyApp(),
    ),
  );
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final TextEditingController controller = TextEditingController(text: 'Collapsed input height');

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: TextField(
            controller: controller,
            autofocus: true,
            decoration: const InputDecoration.collapsed(
              hintText: 'Hint text',
              filled: true,
              border: UnderlineInputBorder(),
            ),
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
}

```

</details>

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Flutter (Channel master, 3.23.0-13.0.pre.277, on macOS 14.2.1 23C71
darwin-arm64, locale fr-FR)
```

</details>

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR sets a minimal height for collapsed `InputDecoration`. 

Before this PR the minimum height was 0. On desktop, due to visual density reducing the height by 8 pixels, it leads to a collapsed text field height being too small to fit the input text vertically.
The following screenshot shows a default collapsed M3 TextField on macOS. On M3 the font style is 16px with a 1.5 height, so the input height is 24. The decoration height is 16 because of the visual density reduction (this results in the border being misplaced, some letters overflowing and the cursor overflowing).

image

After this PR, the minimum height is the input height.

image

## Related Issue

Fixes https://github.com/flutter/flutter/issues/150763

## Tests

Adds 4 tests, updates 2.

### CODE REVIEW

**Comment by bleroux** [comment]
@justinmc I was investigating isCollapsed issues and migrating/adding related tests when I observed https://github.com/flutter/flutter/issues/150763.
Because it is a very small fix, I opted to file the fix with the new migrated tests but if it leads to Google testing failure, I will land the migrated tests separately.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -859,6 +859,8 @@ class _RenderDecoration extends RenderBox with SlottedContainerRenderObjectMixin
     return !decoration.isCollapsed && decoration.border.isOutline;
   }
 
+  Offset get densityOffset => decoration.visualDensity.baseSizeAdjustment;
```
Comment (justinmc) [inline packages/flutter/lib/src/material/input_decorator.dart:862]: Can this be made private?

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -1070,7 +1071,7 @@ class _RenderDecoration extends RenderBox with SlottedContainerRenderObjectMixin
       + densityOffset.dy,
     );
     final double minContainerHeight = decoration.isDense! || decoration.isCollapsed || expands
-      ? 0.0
+      ? inputHeight
```
Comment (justinmc) [inline packages/flutter/lib/src/material/input_decorator.dart:1074]: That looks right, good catch.

File: packages/flutter/test/material/input_decorator_test.dart [hunk 3]
```diff
@@ -6463,6 +6463,120 @@ void main() {
...
+          ),
+        ),
+      );
+
+      expect(getDecoratorRect(tester).size, const Size(800.0, inputHeight));
+      expect(getInputRect(tester).height, inputHeight);
+      expect(getInputRect(tester).top, 0.0);
+      expect(getHintOpacity(tester), 0.0);
+
+      // The hint should appear
```
Comment (justinmc) [inline packages/flutter/test/material/input_decorator_test.dart:6485]: Missing a period here.

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1:

**Pull request merged**